### PR TITLE
Connect to dev environment using the JetBrains SSH connection handler

### PR DESCRIPTION
**Pull request by manodnyab**
## Description

This is an alternate implementation to connecting to Dev Environments. It builds on top of and reuses classes and interfaces from JetBrains's existing SSH connection flow. The previous implementation required a delay of approx 10 secs to prevent deadlocks between various processes that were required to start the IDE Backend. This removes the delay and adds more robustness to the connection logic.

 
## License
I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by justinmk3** [comment]
- screenshot would be helpful
- does this need a changelog?
- a Problem/Solution pair goes a long way to fully answering the purpose of a change. That would answer my next question: how is this different than the current functionality?

File: jetbrains-gateway/src/software/aws/toolkits/jetbrains/gateway/connection/workflow/v2/SsmFilterInputStream.kt [hunk 1]
```diff
@@ -0,0 +1,28 @@
+// Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package software.aws.toolkits.jetbrains.gateway.connection.workflow.v2
+
+import java.io.FilterInputStream
+import java.io.InputStream
+
+// hacky remove trailing newline from SSM output
+class SsmFilterInputStream(source: InputStream) : FilterInputStream(source) {
```
Comment (rli) [inline jetbrains-gateway/src/software/aws/toolkits/jetbrains/gateway/connection/workflow/v2/SsmFilterInputStream.kt:10]: not sure if we still need this

File: jetbrains-gateway/src/software/aws/toolkits/jetbrains/gateway/connection/workflow/v2/CawsHostCommandExecutor.kt [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+import com.intellij.remote.RemoteCredentials
+import com.jetbrains.gateway.ssh.deploy.CommandExecutionResultWithRawStreams
+import com.jetbrains.gateway.ssh.deploy.CommandExecutionStreams
+import com.jetbrains.gateway.ssh.deploy.HostCommandExecutor
+import com.jetbrains.gateway.ssh.deploy.ShellArgument
+import com.jetbrains.gateway.ssh.deploy.impl.SshCommandExecutor
+import com.jetbrains.rd.util.lifetime.Lifetime
+
+class CawsHostCommandExecutor private constructor(private val delegate: HostCommandExecutor) : HostCommandExecutor by delegate {
```
Comment (rli) [inline jetbrains-gateway/src/software/aws/toolkits/jetbrains/gateway/connection/workflow/v2/CawsHostCommandExecutor.kt:14]: if we don't need the hacky input stream filter, we don't need this either

File: jetbrains-gateway/src/software/aws/toolkits/jetbrains/gateway/CawsConnectionProvider.kt [hunk 3]
```diff
@@ -438,7 +439,7 @@ class CawsConnectionProvider : GatewayConnectionProvider {
                 }
             }
 
-            add(StartBackend(gatewayHandle, remoteScriptPath, remoteProjectName, executor, lifetime, envId, isSmallInstance))
+            add(StartBackendV2(lifetime, indicator, envId, remoteProjectName))
```
Comment (rli) [inline jetbrains-gateway/src/software/aws/toolkits/jetbrains/gateway/CawsConnectionProvider.kt:442]: maybe we can rewrite some of the workflow to shave a couple more seconds off

File: jetbrains-gateway/src/software/aws/toolkits/jetbrains/gateway/connection/workflow/v2/SsmFilterInputStream.kt (hunk 1 above)
Comment (manodnyab) [reply jetbrains-gateway/src/software/aws/toolkits/jetbrains/gateway/connection/workflow/v2/SsmFilterInputStream.kt]: Done

File: jetbrains-gateway/src/software/aws/toolkits/jetbrains/gateway/connection/workflow/v2/CawsHostCommandExecutor.kt (hunk 2 above)
Comment (manodnyab) [reply jetbrains-gateway/src/software/aws/toolkits/jetbrains/gateway/connection/workflow/v2/CawsHostCommandExecutor.kt]: Done

File: jetbrains-gateway/src/software/aws/toolkits/jetbrains/gateway/CawsConnectionProvider.kt (hunk 3 above)
Comment (manodnyab) [reply jetbrains-gateway/src/software/aws/toolkits/jetbrains/gateway/CawsConnectionProvider.kt:422]: Planning to address it in another PR

**Comment by manodnyab** [comment]
> * screenshot would be helpful
> 
>     * does this need a changelog?
> 
>     * a Problem/Solution pair goes a long way to fully answering the purpose of a change. That would answer my next question: how is this different than the current functionality?

Updated the description for a better explanation of why this change is required: https://github.com/aws/aws-toolkit-jetbrains/pull/3629#issue-1685963745
The UX stays pretty much the same.
This may need a changelog, will add it as part of another PR

**Review by saurabhajmera — APPROVED** [review]

**Pull request merged**